src/Symfony/Component/Cache/Adapter/PdoTagAwareAdapter.php

```diff
@@ -0,0 +1,239 @@
+                // It also requires HOLDLOCK according to http://weblogs.sqlteam.com/dang/archive/2009/01/31/UPSERT-Race-Condition-With-MERGE.aspx
+                $sql = "MERGE INTO $this->tagsTable WITH (HOLDLOCK) USING (SELECT 1 AS dummy) AS src ON ($this->idCol = :id, $this->tagCol = :tagId) ".
+                    "WHEN NOT MATCHED THEN INSERT ($this->idCol, $this->tagCol) VALUES (:id, :tagId) ".
+                    "WHEN MATCHED THEN UPDATE SET $this->idCol = :id, $this->tagCol = :tagId;";
+                break;
+            case 'sqlite' === $driver:
+                $sql = 'INSERT OR REPLACE'.substr($insertSql, 6);
+                break;
+            case 'pgsql' === $driver && version_compare($this->getServerVersion(), '9.5', '>='):
+                $sql = $insertSql." ON CONFLICT ($this->idCol, $this->tagCol) DO UPDATE SET ($this->idCol, $this->tagCol) = (EXCLUDED.$this->idCol, EXCLUDED.$this->tagCol)";
+                break;
+            default:
+                throw new \DomainException(\sprintf('Caching support is currently not implemented for PDO driver "%s".', $driver));
+        }
+
+        foreach ($addTagData as $tagId => $ids) {
+            foreach ($ids as $id) {
+                if (\in_array($id, $failed, true)) {
+                    continue;
+                }
+
+                try {
+                    $stmt = $this->prepareStatementWithFallback($sql, function () {
+                        $this->createTagsTable();
+                    });
+
+                    $stmt->bindParam(':id', $id);
+                    $stmt->bindParam(':tagId', $tagId);
+
+                    $this->executeStatementWithFallback($stmt, function () {
+                        $this->createTagsTable();
+                    });
+                } catch (\PDOException $e) {
+                    $failed[] = $id;
+                }
+            }
+        }
+
+        foreach ($removeTagData as $tagId => $ids) {
+            foreach ($ids as $id) {
+                if (\in_array($id, $failed, true)) {
+                    continue;
+                }
+
+                $sql = "DELETE FROM $this->tagsTable WHERE $this->idCol=:id AND $this->tagCol=:tagId";
+
+                try {
+                    $stmt = $this->prepareStatementWithFallback($sql, function () {
+                        $this->createTagsTable();
+                    });
+
+                    $stmt->bindParam(':id', $id);
+                    $stmt->bindParam(':tagId', $tagId);
+
+                    $this->executeStatementWithFallback($stmt, function () {
+                        $this->createTagsTable();
+                    });
+                } catch (\PDOException $e) {
+                    $failed[] = $id;
+                }
+            }
+        }
+
+        return $failed;
+    }
+
+    protected function doDeleteTagRelations(array $tagData): bool
+    {
+        foreach ($tagData as $tagId => $idList) {
```

Comment:
is there a chance `$idList` could be an empty array at the point ? This would break the query.

Reply:
I don't think there is but I have added a fallback anyway.